src/Rules/ClassForbiddenNameCheck.php

```diff
@@ -0,0 +1,49 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules;
+
+use function sprintf;
+use function str_starts_with;
+
+class ClassForbiddenNameCheck
+{
+
+	private const INTERNAL_CLASS_PREFIXES = [
+		'PHPStan' => '_PHPStan_',
+		'Rector' => 'RectorPrefix',
+		'PHP-Scoper' => '_PhpScoper',
+	];
+
+	/**
+	 * @param ClassNameNodePair[] $pairs
+	 * @return RuleError[]
+	 */
+	public function checkClassNames(array $pairs): array
+	{
+		$errors = [];
+		foreach ($pairs as $pair) {
+			$className = $pair->getClassName();
+
+			$projectName = null;
+			foreach (self::INTERNAL_CLASS_PREFIXES as $project => $prefix) {
+				if (str_starts_with($className, $prefix)) {
+					$projectName = $project;
+					break;
+				}
+			}
+
+			if ($projectName === null) {
+				continue;
+			}
+
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'Internal %s Class cannot be referenced: %s.',
```

Comment:
Please rewrite this as: "Referencing prefixed PHPStan class: _PHPStan_d5242a5\Symfony/Finder"

And add a tip without the prefix: "This is most likely unintentional. Did you mean to type Symfony\Finder?"